Topics/Software_Engineering/Retries.md

```diff
@@ -0,0 +1,86 @@
+metadata about the request itself. For example, there is a `$retryable` field in most of the AWS SDKv2's APIs, 
+the most common client used to make AWS API calls. If this field is set `true`, the server is hinting that the 
+failure was intermittent and that the client should retry. If the field is set to `false`, the server is hinting 
+to the client that the failure is likely going to happen again.
+
+## What are the problems with retries?
+
+Since retries are so simple to implement and elegant, they are usually the first tool that developers reach for 
+when a dependency of theirs has intermittent failures, but how can this go wrong?
+
+Consider a case where 4 distinct software teams each build products that depend on one another, in a chain like:
+```
+A -> B -> C -> D
+```
+
+That is, service A is calling service B's APIs, and so on. Since B's APIs are known to fail occasionally, A has 
+configured an automatic retry count of 3. Underneath the hood, B depends on C. Service A may or may not know this 
+about B. But since C has a flaky API too, B also has a retry count of 3. And the same for C. 
+
+This works fine, and will usually work. If all services are sufficiently scaled up to handle the load they are 
+given, there are no problems.
+
+However, imagine a case where service D is down. Though it is at the end of the chain of dependencies, in theory, 
+the services should be able to stay up despite their dependencies being down. This type of engineering is called 
+fault-tolerance.
+
+The next time that Service A takes a request, it forwards it to B, which forwards it to C, which tries to call D's 
+API, which fails. C then, tries again 3 times before reporting a failure back to B, which also triggers a retry. 
+That means C tries _another 3 times_.
+
+Retries deep into services grow multiplicatively, and a single API call to A has caused
+```
+A: 3
+B: 9
+C: 27
+```
+different API calls to fail. C is handling 27x more load than it is used to, and might start failing itself, further 
+exacerbating the problem. 
+
+That is, D has become a single point of failure for all other services, and even if they don't outright fail, the 
+load on B, C, and D, are highly needlessly increased.
+
+## What can we do about retries?
+
+Clients calling services will nearly always have retries configured. However, internal services should rarely 
+implement retries while calling other internal services, for precisely this reason.
+
+Another technique to get around excessive retries is to utilize more caching. If service C had cached the responses 
+from service D, it's possible that service D going down would have affected the top-level services at all, and 
+everything would have worked as normal. The downside to this approach is that caches are often trick to get right, 
+and sometimes introduce modal behavior in services [1], usually a bad thing.
+
+## So should I retry?
```

Comment:
I liked how you included this section. It'll be useful for future CSC301 students who are thinking of adding retries into their code.

Reply:
You could also mention different strategies of retries (ex. exponential backoff, capped retries, etc.) just to give an idea of different ways retries can be used.

Reply:
Sure! Good idea.